File: src/main/java/org/khun/leetcode/ArrayAndString/ZizagConversion.java
package org.khun.leetcode.ArrayAndString;

import java.util.Arrays;
import java.util.StringJoiner;

public class ZizagConversion {
    public String convert(String s, int numRows){
        if(numRows == 1 || s.length() <= numRows) return s;

        StringBuilder[] rows = new StringBuilder[numRows];
        for(int i =0; i<rows.length; i++){
            rows[i] = new StringBuilder();
        }
        int curRow = 0;
        boolean direction = false;
        for(int i =0; i<s.length(); i++){
            rows[curRow].append(s.charAt(i));
            if(curRow == 0 || curRow == numRows - 1) direction = !direction;
            if (direction) curRow++;
            else curRow--;
        }

        StringBuilder result = new StringBuilder();
        for (StringBuilder row: rows){
            result.append(row);
        }

        if (result.length() == s.length()) return result.toString();
        else return "";
    }

    public static void main(String[] args) {
        ZizagConversion zizagConversion = new ZizagConversion();
        System.out.println(zizagConversion.convert("PAYPALISHIRING", 3));
        System.out.println(zizagConversion.convert("PAYPALISHIRING", 4));
        System.out.println(zizagConversion.convert("A", 1));
    }
}

File: src/main/java/org/khun/leetcode/ArrayAndString/ZizagConversion.kt
package org.khun.leetcode.ArrayAndString

fun convert(s: String, numRows: Int): String {
    if(numRows == 1 || s.length < numRows) return s

    val rows = Array(numRows){StringBuilder()}
    var curRow = 0
    var direction = false

    for(ch in s){
        rows[curRow].append(ch)
        if(curRow == 0 || curRow == numRows-1) direction = !direction
        curRow += if(direction) 1 else -1
    }

    val result = StringBuilder()
    for(row in rows){
        result.append(row)
    }

    return result.toString()
}

fun main() {
    println(convert("PAYPALISHIRING", 3))
    println(convert("PAYPALISHIRING", 4))
    println(convert("AB", 1))
}